### PR TITLE
Fetch DF Batch function

### PR DESCRIPTION
**Pull request by pdet**
This PR implements a fetch_df_chunk in the python API, that returns one chunk of the query result per time.

It also introduces a CMake file for the python API,  used in conjunction with a BUILD_PYTHON_PKG flag, only to be used on development with clion

### CODE REVIEW

File: src/include/duckdb/main/materialized_query_result.hpp [hunk 1]
```diff
@@ -24,6 +24,8 @@ class MaterializedQueryResult : public QueryResult {
 	DUCKDB_API MaterializedQueryResult(string error);
 
 	ChunkCollection collection;
+	//! In case we are streaming the data we keep track of the current chunk
+	idx_t cur_chunk = 0;
```
Comment (Mytherin) [inline src/include/duckdb/main/materialized_query_result.hpp:28]: The materialized chunk also supports `Fetch`, can't you just call that method for both the streaming and the materialized result?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR, this looks exquisite. One minor question:

File: tools/pythonpkg/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+include_directories(src)
+
+find_package(PythonLibs REQUIRED)
+
+include_directories(.eggs/pybind11-2.6.1-py3.6.egg/pybind11/include/)
```
Comment (hannesmuehleisen) [inline tools/pythonpkg/CMakeLists.txt:5]: this include dir is unlikely to work on someone else's computer

Comment (pdet) [reply tools/pythonpkg/CMakeLists.txt]: Good catch, I thought I had it as a find_package already

**Comment by hannesmuehleisen** [comment]
This should resolve #1261

**Comment by Alex-Monahan** [comment]
Hey Folks!

Thanks again for adding this. Unfortunately, I am running into an error when trying to implement it.  Please let me know if there is a workaround for me to use!

How can I tell when I have pulled the final chunk? If I attempt to call fetch_df_chunk one too many times, it kills my Python kernel (Running in a Jupyter Notebook in Python 3.7 on Windows with DuckDB 0.2.5). Even a try/catch block won't work since the kernel is dying. I also thought about trying to check if the final chunk has fewer rows that the vector size, but if a result set is evenly divisible by the vector size, I will still run into this error. Is there a way to check if the cursor is empty other than calling fetch_df_chunk again? Or could it return None or throw an exception in the empty case or something similar?

Thanks!
-Alex

**Comment by Alex-Monahan** [comment]
import pandas
import duckdb

conn = duckdb.connect(':memory:')
duckdb_cursor = conn.cursor()
duckdb_cursor.execute("CREATE table t as select range a from range(3000);")
query = duckdb_cursor.execute("SELECT a FROM t")
cur_chunk = query.fetch_df_chunk()
print(cur_chunk)
cur_chunk = query.fetch_df_chunk()
print(cur_chunk)
cur_chunk = query.fetch_df_chunk()
print(cur_chunk)
cur_chunk = query.fetch_df_chunk()
print(cur_chunk)
#Should be empty by now
try:
    cur_chunk = query.fetch_df_chunk()
    print(cur_chunk)
except exception as err:
    print(err)
finally:
    conn.close()

**Comment by hannesmuehleisen** [comment]
@pdet can you have a look at @Alex-Monahan's issue please?

**Comment by pdet** [comment]
Yes, no problem!

**Comment by Alex-Monahan** [comment]
Thank you for taking a look! I am able to detect when there are no more results to process in the query now. 

However, I'm getting a pretty severe performance penalty with moderate sized result sets with this approach. A query that returns 1 million rows used to take 400ms and now takes 1.5 seconds (~4x as long). I've attached a Jupyter notebook to illustrate the issue with some randomly generated datasets. I think this is a byproduct of Pandas performance in some way, so it may mean that there may be an advantage to specifying a chunk size that is much larger than the vector size. There is also a further 400ms penalty to concatenate the list of Data Frames together (which may also be related to the number of df's to combine). I think we will stick with the fetchdf approach in our production servers for now given the speed advantage. 

I've also included a function that loops through the chunks within the notebook. I have to add a special case to handle when there are deliberately 0 rows returned by a query to differentiate from when all chunks have been pulled (which is workable, it just means that anyone wanting to use this feature will need to consider that).


[DuckDB fetch_df_chunk performance timing.zip](https://github.com/cwida/duckdb/files/6186018/DuckDB.fetch_df_chunk.performance.timing.zip)


Thank you for your help!
-Alex

**Comment by Alex-Monahan** [comment]
Hi Folks! 

I am revisiting this feature again and had a quick question. 

One way to help address the performance issue with Pandas concatenation is to use a much larger vector size so the concatenations happen less frequently. 

What vector sizes are well tested today, and which would you recommend that I use? I was thinking 1024*8 or 1024*16, but would rather use one that is well tested. Are there any cons/performance consequences to using larger vector sizes? We are operating on a server and using 4 threads for each DuckDB request.

Thanks!
-Alex

**Comment by pdet** [comment]
Hey Alex ,

Looking at the code, it seems to me that this function just exposes the streaming API, so the whole engine is still vectorized on our internal vector size (i.e., 1024 right now), in your code, you would have to call the fetch_df_chunk and concatenate the result.

I could also extend this to receive a parameter that indicates the number of vectors you want to output per chunk, then you could call  fetch_df_chunk(8) and it would output 8*1024 elements.

Note that, even if I allow the user to set the number of vectors he desires to output per chunk, this doesn't really change the vector size used in the execution, but just the chunk size that is being streamed.

**Comment by Alex-Monahan** [comment]
Hmm, is the internal vector size not adjustable from the Python client? I did not see a Pragma for it in the code, but I know that many tests use different vector sizes.

My goal was to not ask for any code modifications from you! I was just thinking that the overall vector size was adjustable and I was planning on increasing it.

If it is not adjustable, then it would be helpful to have a parameter for number of chunks to fetch like you described. Either that or a Pragma for the underlying vector size (if you think it would not slow things down too badly or be too complicated to implement)

**Comment by pdet** [comment]
When the internal vector size is changed the whole code must be recompiled, so it not a parameter that is easily adjustable.
The main idea behind testing with different vector sizes is that our tests are sometimes small (i.e., less than 1024 tuples) so testing with vectors that are small can trigger errors in the vectorized engine that wouldn't be triggered with the standard size.

I think what you want is the adjustable output size, which is not hard to implement, so I can have a go at it either later today or tomorrow and will ping you ;-)

**Comment by Alex-Monahan** [comment]
You are correct that I would not like to recompile DuckDB to change that setting! :-)

Thank you for the additional background and for having another look at this!

**Pull request merged**